admin-dev/themes/new-theme/js/pages/order/index.js

```diff
@@ -34,11 +34,30 @@ import SubmitBulkExtension from '../../components/grid/extension/submit-bulk-act
 import BulkActionCheckboxExtension from '../../components/grid/extension/bulk-action-checkbox-extension';
 import FiltersSubmitButtonEnablerExtension
   from '../../components/grid/extension/filters-submit-button-enabler-extension';
-import ModalFormSubmitExtension from "../../components/grid/extension/modal-form-submit-extension";
+import ModalFormSubmitExtension from '../../components/grid/extension/modal-form-submit-extension';
+import PreviewExtension from '../../components/grid/extension/preview-extension';
+import PreviewRenderer from './preview-renderer';
 
 const $ = window.$;
 
 $(() => {
+  function previewRenderer(id) {
+    return new Promise((resolve, reject) => {
+      $.ajax({
+        url: this.previewDataUrl,
+        method: 'GET',
+        dataType: 'html',
```

Comment:
What do you mean? Instead of HTML we should get JSON data and render template from JS?

Reply:
Yes, and because if a module add something in its template and json, he can automatically add data (maybe we should also trigger js events here)

Reply:
Yeah, here are 2 ways I can see:

1. Get HTML from ajax, in that case we could dispatch hook to allow modules to add custom content to final template.

2. Get JSON data from ajax, but in that case, we would need to render template in JS. It could be a bit more difficult for both us and module developers, because 1. We would need to manage translations in JS 2. Complex template rendering in JS would be harder than in Twig 3. It's harder for module developers to deal with JS event and adding custom content in JS than to execute hook when rendering from backend.

What do you think @PierreRambaud?

Reply:
3. Get JSON data from ajax with `{"htmlContent": "blablabl", "myCompanyHook": "Blabalbal"}` and add a `prestashop.emit('onOrderPreviewDataUrlSomething', {eventType: 'blabla', resp: resp}`. More because you maybe wanted to execute or render something elsewhere the BO is rendering the `htmlContent`. So you can keep the rendering with PHP (if this one is too complicated to render with vuejs or jquery), and also trigger a custom event for modules.